clang/lib/Interpreter/CodeCompletion.cpp

```diff
@@ -56,26 +69,135 @@ class ReplCompletionConsumer : public CodeCompleteConsumer {
   std::shared_ptr<GlobalCodeCompletionAllocator> CCAllocator;
   CodeCompletionTUInfo CCTUInfo;
   std::vector<std::string> &Results;
+  ReplCodeCompletion &CC;
+};
+
+class CompletionContextHanndler {
+protected:
+  CodeCompletionContext CCC;
+  std::vector<std::string> &Results;
+
+public:
+  CompletionContextHanndler(CodeCompletionContext CCC,
+                            std::vector<std::string> &Results)
+      : CCC(CCC), Results(Results) {}
+  virtual void handleDeclaration(const CodeCompletionResult &Result) {}
+  virtual void handleKeyword(const CodeCompletionResult &Result) {}
+  virtual void handlePattern(const CodeCompletionResult &Result) {}
+  virtual void handleMacro(const CodeCompletionResult &Result) {}
+};
+
+class DotMemberAccessHandler : public CompletionContextHanndler {
+public:
+  DotMemberAccessHandler(CodeCompletionContext CCC,
+                         std::vector<std::string> &Results)
+      : CompletionContextHanndler(CCC, Results) {}
+  void handleDeclaration(const CodeCompletionResult &Result) override {
+    if (auto *ID = Result.Declaration->getIdentifier()) {
+      if (const auto *Fun = llvm::dyn_cast<CXXMethodDecl>(Result.Declaration)) {
+        if (Fun->getParent()->getCanonicalDecl() ==
+            CCC.getBaseType()->getAsCXXRecordDecl()->getCanonicalDecl()) {
+          LLVM_DEBUG(llvm::dbgs() << "[In HandleCodeCompleteDOT] Name : "
+                                  << ID->getName() << "\n");
+          Results.push_back(ID->getName().str());
+        }
+      }
+    }
+  }
+};
+
+class DefaultAccessHandler : public CompletionContextHanndler {
+private:
+  Sema &S;
+
+public:
+  DefaultAccessHandler(Sema &S, CodeCompletionContext CCC,
+                       std::vector<std::string> &Results)
+      : CompletionContextHanndler(CCC, Results), S(S) {}
+  void handleDeclaration(const CodeCompletionResult &Result) override {
+    auto PreferredType = CCC.getPreferredType();
+    if (!PreferredType.isNull()) {
```

Comment:
We could use a similar approach here:
```suggestion
    if (PreferredType.isNull())
      return;
    ...
```